File: apps/web/lib/providers/adapter-map.ts
/**
 * The self-registration map behind the Provider Registry's factory. See
 * docs/PROVIDER_REGISTRY.md for the full explanation of why this exists
 * instead of a dynamic `import()` of `providers.adapter_class_identifier`
 * (Next.js/webpack cannot resolve an arbitrary runtime string to a module).
 *
 * Each adapter module calls `registerAdapter(identifier, factoryFn)` at
 * import time (see adapters/mock.ts, adapters/plivo.ts, adapters/frejun-teler.ts).
 * apps/web/lib/providers/registry.ts imports every adapter module once (for
 * side effects, so they register themselves) and then looks up
 * `providers.adapter_class_identifier` in this map — it never contains an
 * if/else or switch on a provider name itself.
 *
 * This map is intentionally layer-agnostic: identifiers are namespaced
 * strings like "telephony.plivo" chosen by convention, not a typed enum, so
 * Phase 3's stt./tts./llm. adapters register into the exact same map with
 * zero changes here.
 */

// eslint-disable-next-line @typescript-eslint/no-explicit-any
type AdapterFactory = (config: Record<string, unknown>) => any;

const registry = new Map<string, AdapterFactory>();

export function registerAdapter(identifier: string, factory: AdapterFactory): void {
  if (registry.has(identifier)) {
    // Re-registration happens harmlessly under Next.js dev hot-reload /
    // repeated test imports of the same module; last write wins.
    registry.set(identifier, factory);
    return;
  }
  registry.set(identifier, factory);
}

export function resolveAdapterFactory(identifier: string): AdapterFactory | undefined {
  return registry.get(identifier);
}

export function listRegisteredAdapterIdentifiers(): string[] {
  return Array.from(registry.keys());
}

/** Test-only escape hatch to remove a fake adapter registered by a test. */
export function _unregisterAdapterForTests(identifier: string): void {
  registry.delete(identifier);
}

File: services/voice-gateway/voice_gateway/adapter_map.py
"""The Python side of the self-registration adapter map.

Same contract as apps/web/lib/providers/adapter-map.ts: each adapter module
calls `register_adapter(identifier, factory)` at import time; `registry.py`
imports every built-in adapter module once (for that side effect) and then
does a single dict lookup on `providers.adapter_class_identifier` — never a
branch on provider identity. See docs/PROVIDER_REGISTRY.md.

Identifiers are the SAME namespaced strings ("stt.sarvam", "tts.cartesia",
"llm.gemini_flash", ...) written into the shared `providers` table by
db/migrations/008_stt_tts_llm_providers.sql — the TS and Python runtimes
both resolve against the identical catalog rows, they just each keep their
own in-process registration map (a Python dict here, a TS Map there).
"""

from __future__ import annotations

from collections.abc import Callable
from typing import Any

AdapterFactory = Callable[[dict[str, Any]], Any]

_registry: dict[str, AdapterFactory] = {}


def register_adapter(identifier: str, factory: AdapterFactory) -> None:
    # Re-registration is harmless (repeated test imports of the same
    # module, module reload) — last write wins, matching adapter-map.ts.
    _registry[identifier] = factory


def resolve_adapter_factory(identifier: str) -> AdapterFactory | None:
    return _registry.get(identifier)


def list_registered_adapter_identifiers() -> list[str]:
    return list(_registry.keys())


def _unregister_adapter_for_tests(identifier: str) -> None:
    _registry.pop(identifier, None)
